File: project-{topic}-advanced/src/middleware/RateLimitMiddleware.cpp
```cpp
#include "RateLimitMiddleware.hpp"

RateLimitMiddleware::RateLimitMiddleware(std::shared_ptr<RateLimiter> rate_limiter)
    : rate_limiter_(rate_limiter) {
    if (!rate_limiter_) {
        Logger::log(LogLevel::CRITICAL, "RateLimitMiddleware initialized with null RateLimiter.");
        throw std::runtime_error("RateLimiter dependency is null.");
    }
}

void RateLimitMiddleware::before_handle(crow::request& req, crow::response& res, void* /*ctx*/) {
    // For rate limiting, we'll use the client's IP address
    std::string client_ip = req.remote_ip_address;
    if (client_ip.empty()) {
        client_ip = "unknown_ip"; // Fallback for local testing or if IP is not resolved
    }

    if (!rate_limiter_->allowRequest(client_ip)) {
        // Rate limit exceeded
        Logger::log(LogLevel::WARNING, "Rate limit exceeded for client: " + client_ip + " on URL: " + req.url);
        res = handle_exception(TooManyRequestsException("Too many requests from this IP address. Please try again later."));
        res.end(); // Immediately send response and stop further processing
    }
    // If allowed, just return and let the request proceed to the next middleware or handler.
}

void RateLimitMiddleware::after_handle(crow::request& /*req*/, crow::response& /*res*/, void* /*ctx*/) {
    // No specific action needed after handling for this rate limiter.
}
```

File: project-{topic}-advanced/src/middleware/rate-limit.middleware.ts
```typescript
import { Request, Response, NextFunction } from 'express';
import rateLimit from 'express-rate-limit';
import { AppError, HttpCode } from '../utils/app-error';
import { RATE_LIMIT_WINDOW_MS, RATE_LIMIT_MAX_REQUESTS } from '../config/constants';
import { logger } from '../utils/logger';

export const apiRateLimiter = rateLimit({
  windowMs: RATE_LIMIT_WINDOW_MS, // 15 minutes
  max: RATE_LIMIT_MAX_REQUESTS, // Limit each IP to 100 requests per `window` (here, per 15 minutes)
  message: (req, res) => {
    logger.warn(`Rate limit exceeded for IP: ${req.ip}`);
    return new AppError(
      'Too many requests from this IP, please try again after 15 minutes',
      HttpCode.TOO_MANY_REQUESTS
    ).serializeErrors();
  },
  standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
  legacyHeaders: false, // Disable the `X-RateLimit-*` headers
  keyGenerator: (req) => req.ip, // Use IP address to identify client
});

export const authRateLimiter = rateLimit({
  windowMs: 5 * 60 * 1000, // 5 minutes
  max: 5, // Limit each IP to 5 login attempts per 5 minutes
  message: (req, res) => {
    logger.warn(`Auth rate limit exceeded for IP: ${req.ip}`);
    return new AppError(
      'Too many authentication attempts from this IP, please try again after 5 minutes',
      HttpCode.TOO_MANY_REQUESTS
    ).serializeErrors();
  },
  standardHeaders: true,
  legacyHeaders: false,
});
```